### PR TITLE
Various updates to the store (efficiency, safety, convenience)

### LINKED ISSUE #32: Accelerate sampling in store
The sampling method of the store is very slow, presumably because of inefficient iteration over the model points. Needs to be investigated and potentially fixed.

### LINKED ISSUE #33: Make DirectoryStore & Dataset save to use
Right now calling dataset on the store will write to the store if the wrong priors are specified. I suggest the following changes:
- Introduce `mode = 'r'` for directory store, so that we have a read-only configuration
- Make Dataset fail when points are requested that are not there (rather than adding them to the store).
- Add an additional `add` method to the store, which explicitly adds points for a given prior.
- Add some `check` method that allows to check whether a specific prior is already covered.
- Add some `info` method that provides information about the contained priors and associated number of samples in the store
- Add the ability to remove the latest prior and points.

### LINKED ISSUE #34: Remove "param_list" argument from DataStore
I think the param_list should be part of the simulator, not the datastore. It is inconvenient that this has to be specified each time the datastore is loaded.

### PR DESCRIPTION
**Pull request by cweniger**
Closes #34 , closes #32 and closes #33

### CODE REVIEW

**Comment by cweniger** [comment]
Two new methods for the store:
- "coverage" tests the coverage of a store for a specific prior
- "add" adds points to the store
- "sample", was there before, but now *only* samples, and returns zero samples in the case the store has not sufficient coverage.

Closes #33

**Comment by bkmi** [comment]
> Closes #34 , closes #32 and closes #33

I have concerns about the solution to #33. See my reply there.

File: swyft/store/store.py [hunk 1]
```diff
@@ -439,27 +454,26 @@ def __init__(
...
-        zdim = group[cls._filesystem.pars].shape[1]
-        return DirectoryStore(params=zdim, path=path)
+# FIXME: Why do we need this?
+#    @classmethod
+#    def load(cls, path: PathType):
+#        """Load existing DirectoryStore."""
+#        zarr_store = zarr.DirectoryStore(path)
+#        group = zarr.group(store=zarr_store)
+#        zdim = group[cls._filesystem.pars].shape[1]
+#        return DirectoryStore(params=zdim, path=path)
```
[the remark this answers is not in the document]
Comment (bkmi) [reply swyft/store/store.py:469]: If you can save the simulator, then it isn't a problem. Adding more things to save seems annoying to me.

Wouldn't this all be solved by removing the parameter names anyhow?

**Review by bkmi — COMMENTED** [review]
Based off the content of the code, I only have this one small comment.

File: swyft/store/store.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cweniger) [reply swyft/store/store.py:469]: The parameterr names were removed. The only place where they are specified is the simulator, and then they are now automatically saved with the corresponding datastores, and subsequent objects.

**Comment by cweniger** [comment]
> Also removing the last prior was not addressed here. I don't really mind, it seems like a features which has not yet be asked for by anyone.

Actually this is something that was requested a few times, because it was possible to "accidentally" add points when running `sample`. This is now addressed through the `add` function, hence it might be less of an issue.

**Pull request merged**